### PR TITLE
Pin the inkwell version to latest stable on master

### PR DESCRIPTION
**Pull request by AdityaAtulTewari**
The current master branch of inkwell seems to change a lot and for compatibility issues it might help to pin the commit.

I was having trouble building this as well because the macro for `llvm_versions` now uses different syntax so I fixed that as well.

Please let me know what steps I can take to help get this merged, I really like the idea of this tool and this allows me to use it for my research.

### CODE REVIEW

**Review by jamesmth — APPROVED** [review]
You are right, we should pin to a specific commit, I overlooked that!

Looks good to me!

**Comment by jamesmth** [comment]
Most MacOS tests failed and some Linux tests failed as well.

For MacOS, it seems that https://github.com/TheDan64/inkwell/commit/5c9f7fcbb0a667f7391b94beb65f1a670ad13221 fixes the issue. For Linux, however, we may have to pin to a an older commit (which would not break on MacOS either).

**Review by jamesmth — CHANGES_REQUESTED** [review]
You are right, we should pin to a specific commit, I overlooked that!

Most MacOS tests failed and some Linux tests failed as well.

For MacOS, it seems that https://github.com/TheDan64/inkwell/commit/5c9f7fcbb0a667f7391b94beb65f1a670ad13221 fixes the issue. For Linux, however, we may have to pin to a an older commit (which would not break on MacOS either).

**Comment by AdityaAtulTewari** [comment]
Cool I'll start the search from there thanks!

**Comment by AdityaAtulTewari** [comment]
is there some sort of docker style tests for this? the different linux instances/would you be opposed to introducing that so I can build the linux builds locally?

**Comment by jamesmth** [comment]
There are no docker images for testing the Linux builds, since I wanted to keep the CI workflow similar between MacOS, Windows and Linux.

If you look at the `.github/workflows/linux.yml` file, you can reproduce the Linux test locally with a few commands. Running the test 4 (which fails) with llvm-12 for instance:
```bash
$ wget -O llvm.tar.gz https://github.com/jamesmth/llvm-project/releases/download/v12.0.1-rust-1.55/llvm-lld-12.0.1-rust-1.55-linux-x86_64.tar.gz
$ tar xf llvm.tar.gz
$ LLVM_SYS_120_PREFIX=$PWD/llvm cargo b --manifest-path tests/plugins/Cargo.toml -p plugin* --no-default-features --features target-x86,llvm12-0
$ llvm/bin/opt --load-pass-plugin="tests/plugins/target/debug/libplugin4.so" --passes="mpass,function(fpass)" tests/test.ll -disable-output
```

**Comment by jamesmth** [comment]
Anyway, I fixed the issue with Linux tests with the following modifications:

Replacing these lines:
https://github.com/jamesmth/llvm-plugin-rs/blob/31cc6f2fc50f99e6b35edf8145a52292fba4a427/llvm-plugin/src/analysis.rs#L74-L81

With:
```rust
let res = crate::get_function_analysis_cached_result(
     self.inner,
     id,
     function.as_value_ref().cast(),
);

if !res.is_null() {
    let res = unsafe { Box::leak(Box::from_raw(res.cast())) };
    Some(res)
} else {
    None
}}
```

And replacing this line:
https://github.com/jamesmth/llvm-plugin-rs/blob/31cc6f2fc50f99e6b35edf8145a52292fba4a427/llvm-plugin/src/analysis.rs#L217

With:
```rust
if !res.is_null() {
    let res = unsafe { Box::leak(Box::from_raw(res.cast())) };
    Some(res)
} else {
    None
}
```

If you apply these changes it should pass the CI tests (hopefully).

File: llvm-plugin/Cargo.toml [hunk 1]
```diff
@@ -53,8 +53,8 @@ target-riscv = ["inkwell/target-riscv"]
 target-all = ["inkwell/target-all"]
 
 [dependencies]
-inkwell = { git = "https://github.com/TheDan64/inkwell", rev = "7684346" }
-inkwell_internals = { git = "https://github.com/TheDan64/inkwell", rev = "7684346" }
+inkwell = { git = "https://github.com/TheDan64/inkwell", rev = "5c9f7fc" }
+inkwell_internals = { git = "https://github.com/TheDan64/inkwell", rev = "5c9f7fc"" }
```
Comment (jamesmth) [inline llvm-plugin/Cargo.toml:57]:
```suggestion
inkwell_internals = { git = "https://github.com/TheDan64/inkwell", rev = "5c9f7fc" }
```

**Review by jamesmth — CHANGES_REQUESTED** [review]

**Review by jamesmth — CHANGES_REQUESTED** [review]
Just apply `rustfmt` and we are good to merge!

**Review by jamesmth — APPROVED** [review]

**Pull request merged**